ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsRange.scala

```diff
@@ -40,25 +49,47 @@ object EventsRange {
     if (startExclusive == Offset.beforeBegin) {
       EmptyEventSeqIdRange.startExclusive
     } else {
-      import com.daml.platform.store.Conversions.OffsetToStatement
-      // This query could be: "select min(event_sequential_id) - 1 from participant_events where event_offset > ${range.startExclusive}"
-      // however there are cases when postgres decides not to use the index. We are forcing the index usage specifying `order by event_offset`
-      SQL"select min(event_sequential_id) from participant_events where event_offset > ${startExclusive} group by event_offset order by event_offset asc limit 1"
-        .as(get[Long](1).singleOpt)(connection)
-        .map(_ - 1L)
-        .getOrElse(EmptyEventSeqIdRange.startExclusive)
+      // TODO(Leo): fuse these two queries
+      readMaxRowId(startExclusive)(connection).getOrElse(
+        readNextRowId(startExclusive)(connection)
+          .map(_ - 1L)
+          .getOrElse(
+            if (isLedgerEmpty(connection))
+              EmptyEventSeqIdRange.startExclusive
+            else
+              throw new IllegalStateException(
+                s"readLowerBound($startExclusive) returned None on non-empty ledger")
```

Comment:
This could also happen if the request has an offset that is "from the future".

Reply:
if it is from the future, `readNextRowId` should take care of this. It does
```
select min(event_sequential_id) from participant_events where event_offset > ${offset}
```
and we subtract `1` from this to get the `startExclusive` row_id. If this query also returns `null` we either got an empty ledger in which case we return `EmptyEventSeqIdRange.startExclusive` or something is very wrong with the above two queries and we throw a runtime exception.